src/outputs/parthenon_hdf5.cpp

```diff
@@ -343,145 +456,10 @@ void PHDF5Output::WriteOutputFileImpl(Mesh *pm, ParameterInput *pin, SimTime *tm
-
-  threshold = Env::get<hsize_t>("H5_alignment_threshold", 0, exists);
-  alignment = Env::get<hsize_t>("H5_alignment_alignment", 8 * MiB, exists2);
-  if (exists || exists2) {
-    PARTHENON_HDF5_CHECK(H5Pset_alignment(acc_file, threshold, alignment));
-  }
-
-  // Defer metadata flush
-  // Default: Disabled
-  bool defer_metadata_flush = Env::get<bool>("H5_defer_metadata_flush", false, exists);
-  if (defer_metadata_flush) {
-    H5AC_cache_config_t cache_config;
-    cache_config.version = H5AC__CURR_CACHE_CONFIG_VERSION;
-    PARTHENON_HDF5_CHECK(H5Pget_mdc_config(acc_file, &cache_config));
-    cache_config.set_initial_size = 1;
-    cache_config.initial_size = 16 * MiB;
-    cache_config.evictions_enabled = 0;
-    cache_config.incr_mode = H5C_incr__off;
-    cache_config.flash_incr_mode = H5C_flash_incr__off;
-    cache_config.decr_mode = H5C_decr__off;
-    PARTHENON_HDF5_CHECK(H5Pset_mdc_config(acc_file, &cache_config));
-  }
-
-  /* create an MPI_INFO object -- on some platforms it is useful to
-     pass some information onto the underlying MPI_File_open call */
-  MPI_Info FILE_INFO_TEMPLATE;
-  PARTHENON_MPI_CHECK(MPI_Info_create(&FILE_INFO_TEMPLATE));
-
-  // Free MPI_Info on error on return or throw
-  struct MPI_InfoDeleter {
-    MPI_Info info;
-    ~MPI_InfoDeleter() { MPI_Info_free(&info); }
-  } delete_info{FILE_INFO_TEMPLATE};
-
-  // Hint specifies the manner in which the file will be accessed until the file is closed
-  const auto access_style =
-      Env::get<std::string>("MPI_access_style", "write_once", exists);
-  PARTHENON_MPI_CHECK(
-      MPI_Info_set(FILE_INFO_TEMPLATE, "access_style", access_style.c_str()));
-
-  // Specifies whether the application may benefit from collective buffering
-  // Default :: collective_buffering is disabled
-  bool collective_buffering = Env::get<bool>("MPI_collective_buffering", false, exists);
-  if (exists) {
-    PARTHENON_MPI_CHECK(MPI_Info_set(FILE_INFO_TEMPLATE, "collective_buffering", "true"));
-    // Specifies the block size to be used for collective buffering file acces
-    const auto cb_block_size =
-        Env::get<std::string>("MPI_cb_block_size", "1048576", exists);
-    PARTHENON_MPI_CHECK(
-        MPI_Info_set(FILE_INFO_TEMPLATE, "cb_block_size", cb_block_size.c_str()));
-    // Specifies the total buffer space that can be used for collective buffering on each
-    // target node, usually a multiple of cb_block_size
-    const auto cb_buffer_size =
-        Env::get<std::string>("MPI_cb_buffer_size", "4194304", exists);
-    PARTHENON_MPI_CHECK(
-        MPI_Info_set(FILE_INFO_TEMPLATE, "cb_buffer_size", cb_buffer_size.c_str()));
-  }
-
-  /* tell the HDF5 library that we want to use MPI-IO to do the writing */
-  PARTHENON_HDF5_CHECK(H5Pset_fapl_mpio(acc_file, MPI_COMM_WORLD, FILE_INFO_TEMPLATE));
-#else
-  hid_t const acc_file = H5P_DEFAULT;
-#endif // ifdef MPI_PARALLEL
+  H5P const acc_file = H5P::FromHIDCheck(HDF5::GenerateFileAccessProps());
```

Comment:
I know this was already in the original code, but why is one `acc_file` `H5P` (for MPI) and the other `hid_t` (for serial)?

Also, now it seems that the function returns `hid_t` and here defined as `H5P`

Reply:
So `H5P` is an automatically garbage-collected wrapper around `hid_t` for property lists. I don't want it garbage collected in `GenerateFileAccessProps`, because I want to return the handle at the end of the function. Think of `GenerateFileAccessProps` as a factory. But it should be garbage collected at the end of `WriteOutputs` so  that we clean up after ourselves. So I take the `hid_t` that I now return, and wrap it in the garbage collected class.